File: day2/advanced-mpi-exercises/jacobi.f90
!
! Jacobi iteration with Blocking Sends and Receives 
! Modified from Archer Training Material (EPCC)
!
! http://www.archer.ac.uk/training/course-material/2016/09/160929_AdvMPI_EPCC/index.php
! 
program jacobi
  use mpi
  implicit none     
  ! Boundary value at the LHS of the pipe
  real(kind=8), parameter :: LEFT = 1.0 
  ! Boundary value at the RHS of the pipe
  real(kind=8), parameter :: RIGHT = 10.0
  ! The convergence to terminate at
  real(kind=8), parameter :: CONVERGENCE_ACCURACY = 1e-4
  ! The maximum number of iterations
  integer, parameter :: MAX_ITERATIONS = 500000
  ! How often to report the norm
  integer, parameter :: REPORT_NORM_PERIOD = 1000

  real(kind=8), dimension(:,:), allocatable :: grid, grid_new, temp
  real(kind=8) :: bnorm, rnorm, norm, tmpnorm
  integer :: i, j, k, ierr, size, myrank, local_nx, nx, ny, requests(4)
  character(len=32) :: arg
  double precision :: start_time

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, myrank, ierr)
  start_time=MPI_Wtime()

  if (myrank ==0 .and. command_argument_count() /= 2) then
     print *, &
          "You must provide two command line arguments, the global size in X and the global size in Y"
     stop
  end if

  call get_command_argument(1, arg)
  read(arg,*) nx
  call get_command_argument(2, arg)
  read(arg,*) ny

  if (myrank==0) then
     print *, "Global size in X=", nx, "Global size in Y=", ny
  end if

  local_nx=nx/size
  if (local_nx * size .lt. nx) then
     if (myrank .lt. nx - local_nx * size) local_nx=local_nx+1
  end if

  allocate(grid(ny, 0:local_nx+1), grid_new(ny, 0:local_nx+1), temp(ny, 0:local_nx+1))

  bnorm=0.0
  tmpnorm=0.0
  rnorm=0.0

  ! intialise values
  if (myrank ==0) then
     grid(:,0)=LEFT
  else
     grid(:,0)=0.0d0
  endif
  if (myrank ==size-1) then
     grid(:,local_nx+1)=RIGHT
  else
     grid(:,local_nx+1)=0.d0
  endif
  grid(:,1:local_nx)=0.d0
  grid_new=grid


  do j=1, local_nx
     do i=1, ny
        tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1)) ** 2)
     end do
  end do
  call mpi_allreduce(tmpnorm, bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
  bnorm=sqrt(bnorm)


  do k=0, MAX_ITERATIONS
     if (myrank .gt. 0) then
        call mpi_send(grid(1,1), ny, MPI_DOUBLE, myrank-1, 0, MPI_COMM_WORLD, requests(1), ierr)
        call mpi_recv(grid(1,0), ny, MPI_DOUBLE, myrank-1, 0, MPI_COMM_WORLD, requests(2), ierr)
     end if
     if (myrank .lt. size-1) then
        call mpi_send(grid(1,local_nx), ny, MPI_DOUBLE, myrank+1, 0, MPI_COMM_WORLD, requests(3), ierr)
        call mpi_recv(grid(1,local_nx+1), ny, MPI_DOUBLE, myrank+1, 0, MPI_COMM_WORLD, requests(4), ierr)
     end if
     tmpnorm=0.0
     do j=1, local_nx
        do i=1, ny
           tmpnorm=tmpnorm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1)) ** 2)
        end do
     end do
     call mpi_allreduce(tmpnorm, rnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
     norm=sqrt(rnorm)/bnorm
     if (norm .lt. CONVERGENCE_ACCURACY) exit

     do j=1, local_nx
        do i=1, ny
           grid_new(i,j)=0.25 * (grid(i-1,j) + grid(i+1,j) + grid(i,j-1) + grid(i,j+1))
        end do
     end do
     temp=grid_new
     grid_new=grid
     grid=temp

     if (mod(k, REPORT_NORM_PERIOD)==0 .and. myrank==0) print *, "Iteration=",k," Relative Norm=",norm
  end do
  if (myrank==0) print *, "Terminated on ",k," iterations, Relative Norm=", norm, " runtime=",MPI_Wtime()-start_time," sec"
  deallocate(grid, grid_new, temp)
  call mpi_finalize(ierr)
end program jacobi

File: day2/advanced-mpi-exercises/jacobi.c
//  
//   Modified from Advanced MPI (Archer Training Material,EPCC)
//  
//   http://www.archer.ac.uk/training/course-material/2016/09/160929_AdvMPI_EPCC/index.php
//   
//
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "mpi.h"

// Boundary value at the top of the domain
#define TOP_VALUES 1.0
// Boundary value at the bottom of the domain
#define BOTTOM_VALUES 10.0
// The maximum number of iterations
#define MAX_ITERATIONS 5000000
// The convergence to terminate at
#define CONVERGENCE_ACCURACY 1e-4
// How often to report the norm
#define REPORT_NORM_PERIOD 1000

int nx, ny;

void initialise(double*, double*, int, int, int);

int main(int argc, char * argv[]) {	
	int size, myrank;
	int ProcTop, ProcBottom;
	int PrevLocalDim;

	MPI_Win win;
	MPI_Aint disp;

	MPI_Init(&argc, &argv);
	MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
	MPI_Comm_size(MPI_COMM_WORLD, &size);

	if (argc != 3) {
		if (myrank==0) fprintf(stderr, "You must provide the size in X and size in Y as arguments to this code\n");
		return -1;
	}
	nx=atoi(argv[1]);
	ny=atoi(argv[2]);

	if (myrank==0) printf("Solving to accuracy of %.0e, global system size is x=%d y=%d\n", CONVERGENCE_ACCURACY, nx, ny);
	int local_nx=nx/size;
	PrevLocalDim = local_nx;
	if (local_nx * size < nx) {
		if (myrank < nx - local_nx * size) local_nx++;
		// if (myrank == nx - local_nx * size) PrevLocalDim++;
	}
	if(myrank == nx%size)
		PrevLocalDim++;

	double * grid = malloc(sizeof(double) * (local_nx + 2) * ny);
	double * grid_new = malloc(sizeof(double) * (local_nx + 2) * ny);
	double * temp = malloc(sizeof(double) * (local_nx + 2) * ny);
	double start_time;

	disp = ny*local_nx*sizeof(double);
	MPI_Win_create(&grid[ny], disp, sizeof(double), MPI_INFO_NULL, MPI_COMM_WORLD, &win);
	// disp = ny*(local_nx + 2)*sizeof(double);
	// MPI_Win_create(grid, disp, sizeof(double), MPI_INFO_NULL, MPI_COMM_WORLD, &win);
	ProcTop = myrank==0 ? MPI_PROC_NULL : myrank-1;
	ProcBottom = myrank==size-1 ? MPI_PROC_NULL : myrank+1;

	initialise(grid, grid_new, local_nx, myrank, size);

	double rnorm=0.0, bnorm=0.0, norm, tmpnorm=0.0;
	MPI_Request requests[]={MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL, MPI_REQUEST_NULL};

	int i,j,k;
	for (i=1;i<=local_nx;i++) {
		for (j=0;j<ny;j++) {		
			tmpnorm=tmpnorm+pow(grid[j+(i*ny)]*4-grid[(j-1) + (i*ny)]-grid[(j+1) + (i*ny)] - grid[j+((i-1)*ny)] - grid[j+((i+1)*ny)], 2);
		}
	}
	MPI_Allreduce(&tmpnorm, &bnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
	bnorm=sqrt(bnorm);

	start_time=MPI_Wtime();
	for (k=0;k<MAX_ITERATIONS;k++) {
 
		/*
		if (myrank > 0) {
			MPI_Isend(&grid[ny], ny, MPI_DOUBLE, myrank-1, 0, MPI_COMM_WORLD, &requests[0]);
			MPI_Irecv(&grid[0], ny, MPI_DOUBLE, myrank-1, 0, MPI_COMM_WORLD, &requests[1]);
		}
		if (myrank < size-1) {
			MPI_Isend(&grid[local_nx * ny], ny, MPI_DOUBLE, myrank+1, 0, MPI_COMM_WORLD, &requests[2]);
			MPI_Irecv(&grid[(local_nx+1) * ny], ny, MPI_DOUBLE, myrank+1, 0, MPI_COMM_WORLD, &requests[3]);
		}
		MPI_Waitall(4, requests, MPI_STATUSES_IGNORE);
		*/

		MPI_Win_fence(MPI_MODE_NOPRECEDE, win);
		// MPI_Win_lock(MPI_LOCK_SHARED, ProcTop, MPI_MODE_NOCHECK, win);
		disp = (PrevLocalDim-1)*ny;
		MPI_Get(&grid[0],ny,MPI_DOUBLE,ProcTop,disp,ny,MPI_DOUBLE,win);
		// MPI_Win_unlock(ProcTop, win);

		// MPI_Win_lock(MPI_LOCK_SHARED, ProcBottom, MPI_MODE_NOCHECK, win);
		disp = 0;
		MPI_Get(&grid[(local_nx+1)*ny],ny,MPI_DOUBLE,ProcBottom,disp,ny,MPI_DOUBLE,win);
		// MPI_Win_unlock(ProcBottom, win);
		MPI_Win_fence(MPI_MODE_NOSUCCEED, win);

		tmpnorm=0.0;
		for (i=1;i<=local_nx;i++) {
			for (j=0;j<ny;j++) {
				tmpnorm=tmpnorm+pow(grid[j+(i*ny)]*4-grid[(j-1) + (i*ny)]-grid[(j+1) + (i*ny)] - grid[j+((i-1)*ny)] - grid[j+((i+1)*ny)], 2);
			}
		}
		MPI_Allreduce(&tmpnorm, &rnorm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
		norm=sqrt(rnorm)/bnorm;
		if (norm < CONVERGENCE_ACCURACY) break;		
		for (i=1;i<=local_nx;i++) {
			for (j=0;j<ny;j++) {
				grid_new[j+(i*ny)]=0.25 * (grid[(j-1) + (i*ny)]+grid[(j+1) + (i*ny)] + grid[j+ ((i+1)*ny)] + grid[j+ ((i-1)*ny)]);
			}
		}
		memcpy(temp, grid_new, sizeof(double) * (local_nx + 2) * ny);
		memcpy(grid_new, grid, sizeof(double) * (local_nx + 2) * ny);
		memcpy(grid, temp, sizeof(double) * (local_nx + 2) * ny);

		if (k % REPORT_NORM_PERIOD == 0 && myrank==0) printf("Iteration= %d Relative Norm=%e\n", k, norm);
	}
	if (myrank==0) printf("\nTerminated on %d iterations, Relative Norm=%e, Total time=%e seconds\n", k, norm,
			MPI_Wtime() - start_time);
	free(grid);
	free(grid_new);
	free(temp);
	MPI_Finalize();
	return 0;
}

/**
 * Initialises the arrays, such that grid contains the boundary conditions at the start and end points and all other
 * points are zero. grid_new is set to equal grid
 */
void initialise(double * grid, double * grid_new, int local_nx, int myrank, int size) {
	int i, j;
	for (j=0;j<ny;j++) {
		grid_new[j]=grid[j]=myrank==0 ? TOP_VALUES: 0;
	}
	for (j=0;j<ny;j++) {
		grid_new[j+(ny*(local_nx+1))]=grid[j+(ny*(local_nx+1))]=myrank==size ? BOTTOM_VALUES: 0;		
	}
	for (i=1;i<=local_nx;i++) {
		for (j=0;j<ny;j++) {
			grid_new[j+(ny*i)]=grid[j+(ny*i)]=0;
		}
	}	
}
